libp2p/crypto/chacha20poly1305.nim

```diff
@@ -68,11 +70,11 @@ proc encrypt*(_: type[ChaChaPoly],
   ourPoly1305CtmulRun(
     unsafeAddr key[0],
     unsafeAddr nonce[0],
-    addr data[0],
+    baseAddr(data),
     data.len,
     ad,
```

Comment:
I'm not a fan of `baseAddr` the way it's currently implemented - it ventures into undefined behavior territory on len == 0

Reply:
The `cast[pointer]` mechanism it uses seems to be the only non-`if`-branchy alternative to the just-doesn't-work-with-`checks` `addr a[0]` version. Nim doesn't appear to supply anything else.

Reply:
yeah - my general take is that whatever `if` there is doesn't materially matter for performance (it'll likely get optimised away), but it gives the code a well-defined behavior (say .. nil) at least that we can recognize in call stacks and crash dumps, should we ever have to debug mess like this

Reply:
It's still probably better to centralize this in `baseAddr`, then. That way one can hope/measure one's intuitions about the `if` not mattering (but it adds two checks: one or two for the `if` itself [because of the codegen patterns for `if`/`else`], and one for element-0-array-access, whereas it's currently a no-op in C codegen, almost) and change everything to be safer/less UB in a centralized way if it makes sense.

Having `if foo.len > 0: addr foo[0] else: nil` everywhere is ugly, even if in codegen terms that's what one wants.

Reply:
> better to centralize this in

works for me